Magento/Sniffs/Classes/ConstructorProxyInterceptorSniff.php

```diff
@@ -0,0 +1,267 @@
+<?php
+/**
+ * Copyright © Magento. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Sniffs\Classes;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+/**
+ * Detects explicit request of proxies and interceptors in constructors
+ *
+ * Search is in variable names and namespaces, including indirect namespaces from use statements
+ */
+class ConstructorProxyInterceptorSniff implements Sniff
+{
+    const CONSTRUCT_METHOD_NAME = '__construct';
+
+    /**
+     * String representation of warning.
+     *
+     * @var string
+     */
+    protected $warningMessage = 'Proxies and interceptors MUST never be explicitly requested in constructors.';
```

Comment:
It would be better to split this sniff into two sniffs to comply with single responsibility principle.

Reply:
I have concerns regarding your comment:
1. This sniff covers one statement from Technical Guidelines and thats why it sounds logical to me to have this one rule.
2. Splitting into two sniffs doesn't sound good in performance prospective (make same check twice).
3. On the other hand, making it more generic sounds good to me.  The name can be changed to `DiscouragedDependenciesSniff` (or something like this) and `$warnNames` property can be public and will be replaceable via `ruleset.xml`.

Reply:
> 3. On the other hand, making it more generic sounds good to me. The name can be changed to DiscouragedDependenciesSniff (or something like this) and $warnNames property can be public and will be replaceable via ruleset.xml.

I do like the sound of the latter - I didn't realise that this could be done. Nice :)